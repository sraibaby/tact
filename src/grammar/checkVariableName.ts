import { throwCompilationError } from "../errors";
import { AbstractSrcInfo } from "./src-info";

export function checkVariableName(name: string, loc: AbstractSrcInfo) {
    if (name.startsWith("__gen")) {
        throwCompilationError(`Variable name cannot start with "__gen"`, loc);
    }
    if (name.startsWith("__tact")) {
        throwCompilationError(`Variable name cannot start with "__tact"`, loc);
    }
}
