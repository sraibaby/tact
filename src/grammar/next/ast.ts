/**
 * AST node constructors are not just constructors: they also generate ids
 * We have this file so that the "current id" state would not be stored globally
 */

import { Location } from "@langtools/pgen/runtime";
import * as A from "../ast";
import { AbstractSrcInfo } from "../src-info";

export const getAstSchema = (
    toSrcInfo: (location: Location) => AbstractSrcInfo,
    nextId: number = 1,
) => {
    const createAstNode = <T>(src: Omit<T, "id">): T => {
        return Object.freeze(Object.assign({ id: nextId++ }, src)) as T;
    };

    const cloneAstNode = <T extends { id: number }>(src: T): T => {
        return { ...src, id: nextId++ };
    };

    const constructors = {
        Module: (imports: A.AstImport[], items: A.AstModuleItem[]): A.AstModule => createAstNode<A.AstModule>({ kind: "module", imports, items, }),
        Import: (path: A.AstString, loc: Location): A.AstImport => createAstNode<A.AstImport>({ kind: "import", path, loc: toSrcInfo(loc), }),
        PrimitiveTypeDecl: (name: A.AstId, loc: Location): A.AstPrimitiveTypeDecl => createAstNode<A.AstPrimitiveTypeDecl>({ kind: "primitive_type_decl", name, loc: toSrcInfo(loc), }),
        FunctionDef: (attributes: A.AstFunctionAttribute[], name: A.AstId, retType: A.AstType | null, params: A.AstTypedParameter[], statements: A.AstStatement[], loc: Location): A.AstFunctionDef => createAstNode<A.AstFunctionDef>({ kind: "function_def", attributes, name, return: retType, params, statements, loc: toSrcInfo(loc), }),
        AsmFunctionDef: (shuffle: A.AstAsmShuffle, attributes: A.AstFunctionAttribute[], name: A.AstId, retType: A.AstType | null, params: A.AstTypedParameter[], instructions: readonly A.AstAsmInstruction[], loc: Location): A.AstAsmFunctionDef => createAstNode<A.AstAsmFunctionDef>({ kind: "asm_function_def", shuffle, attributes, name, return: retType, params, instructions, loc: toSrcInfo(loc), }),
        FunctionDecl: (attributes: A.AstFunctionAttribute[], name: A.AstId, retType: A.AstType | null, params: A.AstTypedParameter[], loc: Location): A.AstFunctionDecl => createAstNode<A.AstFunctionDecl>({ kind: "function_decl", attributes, name, return: retType, params, loc: toSrcInfo(loc), }),
        NativeFunctionDecl: (attributes: A.AstFunctionAttribute[], name: A.AstId, nativeName: A.AstFuncId, params: A.AstTypedParameter[], retType: A.AstType | null, loc: Location): A.AstNativeFunctionDecl => createAstNode<A.AstNativeFunctionDecl>({ kind: "native_function_decl", attributes, name, nativeName, params, return: retType, loc: toSrcInfo(loc), }),
        ConstantDef: (attributes: A.AstConstantAttribute[], name: A.AstId, type: A.AstType, initializer: A.AstExpression, loc: Location): A.AstConstantDef => createAstNode<A.AstConstantDef>({ kind: "constant_def", attributes, name, type, initializer, loc: toSrcInfo(loc), }),
        ConstantDecl: (attributes: A.AstConstantAttribute[], name: A.AstId, type: A.AstType, loc: Location): A.AstConstantDecl => createAstNode<A.AstConstantDecl>({ kind: "constant_decl", attributes, name, type, loc: toSrcInfo(loc), }),
        StructDecl: (name: A.AstId, fields: A.AstFieldDecl[], loc: Location): A.AstStructDecl => createAstNode<A.AstStructDecl>({ kind: "struct_decl", name, fields, loc: toSrcInfo(loc), }),
        MessageDecl: (name: A.AstId, opcode: A.AstNumber | null, fields: A.AstFieldDecl[], loc: Location): A.AstMessageDecl => createAstNode<A.AstMessageDecl>({ kind: "message_decl", name, opcode, fields, loc: toSrcInfo(loc), }),
        Contract: (name: A.AstId, traits: A.AstId[], attributes: A.AstContractAttribute[], declarations: A.AstContractDeclaration[], loc: Location): A.AstContract => createAstNode<A.AstContract>({ kind: "contract", name, traits, attributes, declarations, loc: toSrcInfo(loc), }),
        Trait: (name: A.AstId, traits: A.AstId[], attributes: A.AstContractAttribute[], declarations: A.AstTraitDeclaration[], loc: Location): A.AstTrait => createAstNode<A.AstTrait>({ kind: "trait", name, traits, attributes, declarations, loc: toSrcInfo(loc), }),
        FieldDecl: (name: A.AstId, type: A.AstType, initializer: A.AstExpression | null, as: A.AstId | null, loc: Location): A.AstFieldDecl => createAstNode<A.AstFieldDecl>({ kind: "field_decl", name, type, initializer, as, loc: toSrcInfo(loc), }),
        Receiver: (selector: A.AstReceiverKind, statements: A.AstStatement[], loc: Location): A.AstReceiver => createAstNode<A.AstReceiver>({ kind: "receiver", selector, statements, loc: toSrcInfo(loc), }),
        ContractInit: (params: A.AstTypedParameter[], statements: A.AstStatement[], loc: Location): A.AstContractInit => createAstNode<A.AstContractInit>({ kind: "contract_init", params, statements, loc: toSrcInfo(loc), }),
        StatementLet: (name: A.AstId, type: A.AstType | null, expression: A.AstExpression, loc: Location): A.AstStatementLet => createAstNode<A.AstStatementLet>({ kind: "statement_let", name, type, expression, loc: toSrcInfo(loc), }),
        StatementReturn: (expression: A.AstExpression | null, loc: Location): A.AstStatementReturn => createAstNode<A.AstStatementReturn>({ kind: "statement_return", expression, loc: toSrcInfo(loc), }),
        StatementExpression: (expression: A.AstExpression, loc: Location): A.AstStatementExpression => createAstNode<A.AstStatementExpression>({ kind: "statement_expression", expression, loc: toSrcInfo(loc), }),
        StatementAssign: (path: A.AstExpression, expression: A.AstExpression, loc: Location): A.AstStatementAssign => createAstNode<A.AstStatementAssign>({ kind: "statement_assign", path, expression, loc: toSrcInfo(loc), }),
        StatementAugmentedAssign: (op: A.AstAugmentedAssignOperation, path: A.AstExpression, expression: A.AstExpression, loc: Location): A.AstStatementAugmentedAssign => createAstNode<A.AstStatementAugmentedAssign>({ kind: "statement_augmentedassign", op, path, expression, loc: toSrcInfo(loc), }),
        Condition: (condition: A.AstExpression, trueStatements: A.AstStatement[], falseStatements: A.AstStatement[] | null, elseif: A.AstCondition | null, loc: Location): A.AstCondition => createAstNode<A.AstCondition>({ kind: "statement_condition", condition, trueStatements, falseStatements, elseif, loc: toSrcInfo(loc), }),
        StatementWhile: (condition: A.AstExpression, statements: A.AstStatement[], loc: Location): A.AstStatementWhile => createAstNode<A.AstStatementWhile>({ kind: "statement_while", condition, statements, loc: toSrcInfo(loc), }),
        StatementUntil: (condition: A.AstExpression, statements: A.AstStatement[], loc: Location): A.AstStatementUntil => createAstNode<A.AstStatementUntil>({ kind: "statement_until", condition, statements, loc: toSrcInfo(loc), }),
        StatementRepeat: (iterations: A.AstExpression, statements: A.AstStatement[], loc: Location): A.AstStatementRepeat => createAstNode<A.AstStatementRepeat>({ kind: "statement_repeat", iterations, statements, loc: toSrcInfo(loc), }),
        StatementTry: (statements: A.AstStatement[], loc: Location): A.AstStatementTry => createAstNode<A.AstStatementTry>({ kind: "statement_try", statements, loc: toSrcInfo(loc), }),
        StatementTryCatch: (statements: A.AstStatement[], catchName: A.AstId, catchStatements: A.AstStatement[], loc: Location): A.AstStatementTryCatch => createAstNode<A.AstStatementTryCatch>({ kind: "statement_try_catch", statements, catchName, catchStatements, loc: toSrcInfo(loc), }),
        StatementForEach: (keyName: A.AstId, valueName: A.AstId, map: A.AstExpression, statements: A.AstStatement[], loc: Location): A.AstStatementForEach => createAstNode<A.AstStatementForEach>({ kind: "statement_foreach", keyName, valueName, map, statements, loc: toSrcInfo(loc), }),
        TypeId: (text: string, loc: Location): A.AstTypeId => createAstNode<A.AstTypeId>({ kind: "type_id", text, loc: toSrcInfo(loc), }),
        OptionalType: (typeArg: A.AstType, loc: Location): A.AstOptionalType => createAstNode<A.AstOptionalType>({ kind: "optional_type", typeArg, loc: toSrcInfo(loc), }),
        MapType: (keyType: A.AstTypeId, keyStorageType: A.AstId | null, valueType: A.AstTypeId, valueStorageType: A.AstId | null, loc: Location): A.AstMapType => createAstNode<A.AstMapType>({ kind: "map_type", keyType, keyStorageType, valueType, valueStorageType, loc: toSrcInfo(loc), }),
        BouncedMessageType: (messageType: A.AstTypeId, loc: Location): A.AstBouncedMessageType => createAstNode<A.AstBouncedMessageType>({ kind: "bounced_message_type", messageType, loc: toSrcInfo(loc), }),
        OpBinary: (op: A.AstBinaryOperation, left: A.AstExpression, right: A.AstExpression, loc: Location): A.AstOpBinary => createAstNode<A.AstOpBinary>({ kind: "op_binary", op, left, right, loc: toSrcInfo(loc), }),
        OpUnary: (op: A.AstUnaryOperation, operand: A.AstExpression, loc: Location): A.AstOpUnary => createAstNode<A.AstOpUnary>({ kind: "op_unary", op, operand, loc: toSrcInfo(loc), }),
        FieldAccess: (aggregate: A.AstExpression, field: A.AstId, loc: Location): A.AstFieldAccess => createAstNode<A.AstFieldAccess>({ kind: "field_access", aggregate, field, loc: toSrcInfo(loc), }),
        MethodCall: (self: A.AstExpression, method: A.AstId, args: A.AstExpression[], loc: Location): A.AstMethodCall => createAstNode<A.AstMethodCall>({ kind: "method_call", self, method, args, loc: toSrcInfo(loc), }),
        StaticCall: (funcId: A.AstId, args: A.AstExpression[], loc: Location): A.AstStaticCall => createAstNode<A.AstStaticCall>({ kind: "static_call", function: funcId, args, loc: toSrcInfo(loc), }),
        StructInstance: (type: A.AstId, args: A.AstStructFieldInitializer[], loc: Location): A.AstStructInstance => createAstNode<A.AstStructInstance>({ kind: "struct_instance", type, args, loc: toSrcInfo(loc), }),
        StructFieldInitializer: (field: A.AstId, initializer: A.AstExpression, loc: Location): A.AstStructFieldInitializer => createAstNode<A.AstStructFieldInitializer>({ kind: "struct_field_initializer", field, initializer, loc: toSrcInfo(loc), }),
        InitOf: (contract: A.AstId, args: A.AstExpression[], loc: Location): A.AstInitOf => createAstNode<A.AstInitOf>({ kind: "init_of", contract, args, loc: toSrcInfo(loc), }),
        Conditional: (condition: A.AstExpression, thenBranch: A.AstExpression, elseBranch: A.AstExpression, loc: Location): A.AstConditional => createAstNode<A.AstConditional>({ kind: "conditional", condition, thenBranch, elseBranch, loc: toSrcInfo(loc), }),
        Id: (text: string, loc: Location): A.AstId => createAstNode<A.AstId>({ kind: "id", text, loc: toSrcInfo(loc), }),
        FuncId: (text: string, loc: Location): A.AstFuncId => createAstNode<A.AstFuncId>({ kind: "func_id", text, loc: toSrcInfo(loc), }),
        Null: (loc: Location): A.AstNull => createAstNode<A.AstNull>({ kind: "null", loc: toSrcInfo(loc), }),
        String: (value: string, loc: Location): A.AstString => createAstNode<A.AstString>({ kind: "string", value, loc: toSrcInfo(loc), }),
        Boolean: (value: boolean, loc: Location): A.AstBoolean => createAstNode<A.AstBoolean>({ kind: "boolean", value, loc: toSrcInfo(loc), }),
        Number: (base: A.AstNumberBase, value: bigint, loc: Location): A.AstNumber => createAstNode<A.AstNumber>({ kind: "number", base, value, loc: toSrcInfo(loc), }),
        ContractAttribute: (name: A.AstString, loc: Location): A.AstContractAttribute => createAstNode<A.AstContractAttribute>({ type: "interface", name, loc: toSrcInfo(loc), }),
        FunctionAttribute: (type: A.AstFunctionAttributeName, loc: Location): A.AstFunctionAttribute => ({ type, loc: toSrcInfo(loc), }),
        ConstantAttribute: (type: A.AstConstantAttributeName, loc: Location): A.AstConstantAttribute => ({ type, loc: toSrcInfo(loc), }),
        TypedParameter: (name: A.AstId, type: A.AstType, loc: Location): A.AstTypedParameter => createAstNode<A.AstTypedParameter>({ kind: "typed_parameter", name, type, loc: toSrcInfo(loc), }),
    };

    return {
        cloneAstNode,
        constructors,
    };
};

/**
 * List of all constructors for AST nodes
 */
export type AstSchema = ReturnType<typeof getAstSchema>["constructors"];