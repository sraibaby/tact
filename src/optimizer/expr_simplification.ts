import { CompilerContext } from "../context";
import { TactConstEvalError, throwInternalCompilerError } from "../errors";
import {
    AstCondition,
    AstContractDeclaration,
    AstExpression,
    AstStatement,
    AstTraitDeclaration,
    AstTypeDecl,
    AstValue,
    cloneAstNode,
    idText,
    isValue,
    SrcInfo,
} from "../grammar/ast";
import { Interpreter } from "../interpreter";
import {
    getAllStaticConstants,
    getAllStaticFunctions,
    getAllTypes,
    getType,
    replaceStaticConstants,
    replaceStaticFunctions,
    replaceTypes,
} from "../types/resolveDescriptors";
import { getExpType, registerExpType } from "../types/resolveExpression";
import {
    ConstantDescription,
    FieldDescription,
    FunctionDescription,
    InitDescription,
    ReceiverDescription,
    TypeDescription,
    TypeRef,
    Value,
} from "../types/types";
import { makeValueExpression } from "./util";

export function simplify_expressions(ctx: CompilerContext): CompilerContext {
    // The interpreter in charge of simplifying expressions
    const interpreter = new Interpreter(ctx);

    // Traverse the program and attempt to evaluate every expression

    // Process functions
    const newStaticFunctions: Map<string, FunctionDescription> = new Map();

    for (const f of getAllStaticFunctions(ctx)) {
        if (f.ast.kind === "function_def") {
            const statementsResult = process_statements(
                f.ast.statements,
                ctx,
                interpreter,
            );
            const newStatements = statementsResult.stmts;
            ctx = statementsResult.ctx;
            const newFunctionCode = cloneAstNode({
                ...f.ast,
                statements: newStatements,
            });
            newStaticFunctions.set(f.name, { ...f, ast: newFunctionCode });
        } else {
            // The rest of kinds do not have explicit Tact expressions, so just copy the current function description
            newStaticFunctions.set(f.name, f);
        }
    }
    ctx = replaceStaticFunctions(ctx, newStaticFunctions);

    // Process all static constants
    const newStaticConstants: Map<string, ConstantDescription> = new Map();

    for (const c of getAllStaticConstants(ctx)) {
        if (c.ast.kind === "constant_def") {
            const expressionResult = process_expression(
                c.ast.initializer,
                ctx,
                interpreter,
            );
            const newInitializer = expressionResult.expr;
            ctx = expressionResult.ctx;
            const newConstantCode = cloneAstNode({
                ...c.ast,
                initializer: newInitializer,
            });
            newStaticConstants.set(c.name, { ...c, ast: newConstantCode });
        } else {
            // The rest of kinds do not have explicit Tact expressions, so just copy the current description
            newStaticConstants.set(c.name, c);
        }
    }
    ctx = replaceStaticConstants(ctx, newStaticConstants);

    // Process all types

    /**
     * By calling the function getAllTypes on the context object "ctx", one gets an array of TypeDescriptions.
     * Each TypeDescription stores the type declarations in two different ways:
     * - Directly in the TypeDescription object there are fields, constants, and method
     *   declarations. However, these declarations are "coalesced" in the following sense:
     *     If the TypeDescription is a contract, it will contain copies of methods, constants and fields of traits that the
     *     contract inherits from. Similarly, each trait will have declarations of other traits
     *     that the trait inherits from.
     *
     *   For example, if we look into the "functions" property of the TypeDescription object of a contract
     *   we will find functions defined in BaseTrait.
     *
     * - Indirectly in the "ast" property of the TypeDescription. Contrary to the previous case,
     *   the fields, constants and methods in the ast property are NOT coalesced. This means, for example,
     *   that the methods in a TypeDescription's ast of a contract will be methods that are actually
     *   declared in the contract and not in some trait that the contract inherits from.
     *
     * The above means that we will need to process the properties in TypeDescription first,
     * and then use those properties to build the AST (carefully ensuring that only fields, constants and methods
     * that were in the original AST, remain in the new AST).
     */
    const newTypes: Map<string, TypeDescription> = new Map();

    for (const t of getAllTypes(ctx)) {
        let newInitializer: InitDescription | null = null;

        // Process init
        if (t.init) {
            const statementsResult = process_statements(
                t.init.ast.statements,
                ctx,
                interpreter,
            );
            const newStatements = statementsResult.stmts;
            ctx = statementsResult.ctx;
            const newInitCode = cloneAstNode({
                ...t.init.ast,
                statements: newStatements,
            });
            newInitializer = { ...t.init, ast: newInitCode };
        }

        // Process constants
        const newConstants: ConstantDescription[] = [];

        // This map will be used to quickly recover the new definitions when
        // building the AST later
        const newConstantsMap: Map<string, ConstantDescription> = new Map();

        for (const c of t.constants) {
            if (c.ast.kind === "constant_def") {
                const expressionResult = process_expression(
                    c.ast.initializer,
                    ctx,
                    interpreter,
                );
                const newInitializer = expressionResult.expr;
                ctx = expressionResult.ctx;
                const newConstantCode = cloneAstNode({
                    ...c.ast,
                    initializer: newInitializer,
                });
                const newConstantDescription = { ...c, ast: newConstantCode };
                newConstants.push(newConstantDescription);
                newConstantsMap.set(c.name, newConstantDescription);
            } else {
                // The rest of kinds do not have explicit Tact expressions, so just copy the current description
                newConstants.push(c);
                newConstantsMap.set(c.name, c);
            }
        }

        // Process fields
        const newFields: FieldDescription[] = [];

        // This map will be used to quickly recover the new definitions when
        // building the AST later
        const newFieldsMap: Map<string, FieldDescription> = new Map();

        for (const f of t.fields) {
            if (f.ast.initializer !== null) {
                const expressionResult = process_expression(
                    f.ast.initializer,
                    ctx,
                    interpreter,
                );
                const newInitializer = expressionResult.expr;
                ctx = expressionResult.ctx;
                const newFieldCode = cloneAstNode({
                    ...f.ast,
                    initializer: newInitializer,
                });
                const newFieldDescription = { ...f, ast: newFieldCode };
                newFields.push(newFieldDescription);
                newFieldsMap.set(f.name, newFieldDescription);
            } else {
                // Field without initializer, no expression to simplify inside
                newFields.push(f);
                newFieldsMap.set(f.name, f);
            }
        }

        // Process receivers
        const newReceivers: ReceiverDescription[] = [];

        // This map will be used to quickly recover the new definitions when
        // building the AST later.
        // Since receivers do not have names, I will use their id in their original ast
        // as key.
        const newReceiversMap: Map<number, ReceiverDescription> = new Map();

        for (const r of t.receivers) {
            const statementsResult = process_statements(
                r.ast.statements,
                ctx,
                interpreter,
            );
            const newStatements = statementsResult.stmts;
            ctx = statementsResult.ctx;
            const newReceiverCode = cloneAstNode({
                ...r.ast,
                statements: newStatements,
            });
            const newReceiverDescription = { ...r, ast: newReceiverCode };
            newReceivers.push(newReceiverDescription);
            newReceiversMap.set(r.ast.id, newReceiverDescription);
        }

        // Process methods

        // This is already a map in TypeDescription. This is the reason
        // I did not need a separate map, like in the previous cases.
        const newMethods: Map<string, FunctionDescription> = new Map();

        for (const [name, m] of t.functions) {
            if (m.ast.kind === "function_def") {
                const statementsResult = process_statements(
                    m.ast.statements,
                    ctx,
                    interpreter,
                );
                const newStatements = statementsResult.stmts;
                ctx = statementsResult.ctx;
                const newMethodCode = cloneAstNode({
                    ...m.ast,
                    statements: newStatements,
                });
                newMethods.set(name, { ...m, ast: newMethodCode });
            } else {
                // The rest of kinds do not have explicit Tact expressions, so just copy the current function description
                newMethods.set(name, m);
            }
        }

        // Now, we need to create the new AST, depending on its kind.
        let newAst: AstTypeDecl;

        switch (t.ast.kind) {
            case "primitive_type_decl": {
                newAst = t.ast;
                break;
            }
            case "struct_decl":
            case "message_decl": {
                newAst = cloneAstNode({
                    ...t.ast,
                    fields: newFields.map((f) => f.ast),
                });
                break;
            }
            case "trait": {
                const newDeclarations: AstTraitDeclaration[] = [];

                for (const decl of t.ast.declarations) {
                    switch (decl.kind) {
                        case "asm_function_def":
                        case "function_decl":
                        case "function_def": {
                            const newCode = newMethods.get(idText(decl.name))!
                                .ast as AstTraitDeclaration;
                            newDeclarations.push(newCode);
                            break;
                        }
                        case "constant_decl":
                        case "constant_def": {
                            const newCode = newConstantsMap.get(
                                idText(decl.name),
                            )!.ast;
                            newDeclarations.push(newCode);
                            break;
                        }
                        case "field_decl": {
                            const newCode = newFieldsMap.get(
                                idText(decl.name),
                            )!.ast;
                            newDeclarations.push(newCode);
                            break;
                        }
                        case "receiver": {
                            const newCode = newReceiversMap.get(decl.id)!.ast;
                            newDeclarations.push(newCode);
                            break;
                        }
                    }
                }

                newAst = cloneAstNode({
                    ...t.ast,
                    declarations: newDeclarations,
                });

                break;
            }
            case "contract": {
                const newDeclarations: AstContractDeclaration[] = [];

                for (const decl of t.ast.declarations) {
                    switch (decl.kind) {
                        case "asm_function_def":
                        case "function_def": {
                            const newCode = newMethods.get(idText(decl.name))!
                                .ast as AstContractDeclaration;
                            newDeclarations.push(newCode);
                            break;
                        }
                        case "constant_def": {
                            const newCode = newConstantsMap.get(
                                idText(decl.name),
                            )!.ast as AstContractDeclaration;
                            newDeclarations.push(newCode);
                            break;
                        }
                        case "field_decl": {
                            const newCode = newFieldsMap.get(
                                idText(decl.name),
                            )!.ast;
                            newDeclarations.push(newCode);
                            break;
                        }
                        case "receiver": {
                            const newCode = newReceiversMap.get(decl.id)!.ast;
                            newDeclarations.push(newCode);
                            break;
                        }
                        case "contract_init":
                            newDeclarations.push(newInitializer!.ast);
                            break;
                    }
                }

                newAst = cloneAstNode({
                    ...t.ast,
                    declarations: newDeclarations,
                });

                break;
            }
        }

        newTypes.set(t.name, {
            ...t,
            ast: newAst,
            init: newInitializer,
            constants: newConstants,
            fields: newFields,
            functions: newMethods,
            receivers: newReceivers,
        });
    }
    ctx = replaceTypes(ctx, newTypes);

    return ctx;
}

function process_statements(
    statements: AstStatement[],
    ctx: CompilerContext,
    interpreter: Interpreter,
): { stmts: AstStatement[]; ctx: CompilerContext } {
    const newStatements: AstStatement[] = [];

    for (const stmt of statements) {
        const result = process_statement(stmt, ctx, interpreter);
        newStatements.push(result.stmt);
        ctx = result.ctx;
    }

    return { stmts: newStatements, ctx: ctx };
}

function process_statement(
    stmt: AstStatement,
    ctx: CompilerContext,
    interpreter: Interpreter,
): { stmt: AstStatement; ctx: CompilerContext } {
    switch (stmt.kind) {
        case "statement_assign":
        case "statement_expression":
        case "statement_let":
        case "statement_destruct":
        case "statement_augmentedassign": {
            const expressionResult = process_expression(
                stmt.expression,
                ctx,
                interpreter,
            );
            const new_expr = expressionResult.expr;
            ctx = expressionResult.ctx;

            // Create the replacement node
            return {
                stmt: cloneAstNode({
                    ...stmt,
                    expression: new_expr,
                }),
                ctx: ctx,
            };
        }
        case "statement_return": {
            if (stmt.expression !== null) {
                const expressionResult = process_expression(
                    stmt.expression,
                    ctx,
                    interpreter,
                );
                const new_expr = expressionResult.expr;
                ctx = expressionResult.ctx;

                // Create the replacement node
                return {
                    stmt: cloneAstNode({
                        ...stmt,
                        expression: new_expr,
                    }),
                    ctx: ctx,
                };
            }
            return {
                stmt: stmt,
                ctx: ctx,
            };
        }
        case "statement_condition": {
            const expressionResult = process_expression(
                stmt.condition,
                ctx,
                interpreter,
            );
            const newCondition = expressionResult.expr;
            ctx = expressionResult.ctx;

            const trueStatementsResult = process_statements(
                stmt.trueStatements,
                ctx,
                interpreter,
            );
            const newTrueStatements = trueStatementsResult.stmts;
            ctx = trueStatementsResult.ctx;

            let newFalseStatements: AstStatement[] | null = null;
            if (stmt.falseStatements !== null) {
                const falseStatementsResult = process_statements(
                    stmt.falseStatements,
                    ctx,
                    interpreter,
                );
                newFalseStatements = falseStatementsResult.stmts;
                ctx = falseStatementsResult.ctx;
            }

            let newElseIf: AstCondition | null = null;
            if (stmt.elseif !== null) {
                const elseIfResult = process_statement(
                    stmt.elseif,
                    ctx,
                    interpreter,
                );
                newElseIf = elseIfResult.stmt as AstCondition;
                ctx = elseIfResult.ctx;
            }

            // Create the replacement node
            return {
                stmt: cloneAstNode({
                    ...stmt,
                    condition: newCondition,
                    trueStatements: newTrueStatements,
                    falseStatements: newFalseStatements,
                    elseif: newElseIf,
                }),
                ctx: ctx,
            };
        }
        case "statement_foreach": {
            const expressionResult = process_expression(
                stmt.map,
                ctx,
                interpreter,
            );
            const newMap = expressionResult.expr;
            ctx = expressionResult.ctx;

            const statementsResult = process_statements(
                stmt.statements,
                ctx,
                interpreter,
            );
            const newStatements = statementsResult.stmts;
            ctx = statementsResult.ctx;

            // Create the replacement node
            return {
                stmt: cloneAstNode({
                    ...stmt,
                    map: newMap,
                    statements: newStatements,
                }),
                ctx: ctx,
            };
        }
        case "statement_until":
        case "statement_while": {
            const expressionResult = process_expression(
                stmt.condition,
                ctx,
                interpreter,
            );
            const newCondition = expressionResult.expr;
            ctx = expressionResult.ctx;

            const statementsResult = process_statements(
                stmt.statements,
                ctx,
                interpreter,
            );
            const newStatements = statementsResult.stmts;
            ctx = statementsResult.ctx;

            // Create the replacement node
            return {
                stmt: cloneAstNode({
                    ...stmt,
                    condition: newCondition,
                    statements: newStatements,
                }),
                ctx: ctx,
            };
        }
        case "statement_repeat": {
            const expressionResult = process_expression(
                stmt.iterations,
                ctx,
                interpreter,
            );
            const newIterations = expressionResult.expr;
            ctx = expressionResult.ctx;

            const statementsResult = process_statements(
                stmt.statements,
                ctx,
                interpreter,
            );
            const newStatements = statementsResult.stmts;
            ctx = statementsResult.ctx;

            // Create the replacement node
            return {
                stmt: cloneAstNode({
                    ...stmt,
                    iterations: newIterations,
                    statements: newStatements,
                }),
                ctx: ctx,
            };
        }
        case "statement_try": {
            const statementsResult = process_statements(
                stmt.statements,
                ctx,
                interpreter,
            );
            const newStatements = statementsResult.stmts;
            ctx = statementsResult.ctx;

            // Create the replacement node
            return {
                stmt: cloneAstNode({
                    ...stmt,
                    statements: newStatements,
                }),
                ctx: ctx,
            };
        }
        case "statement_try_catch": {
            const statementsResult = process_statements(
                stmt.statements,
                ctx,
                interpreter,
            );
            const newStatements = statementsResult.stmts;
            ctx = statementsResult.ctx;

            const catchStatementsResult = process_statements(
                stmt.catchStatements,
                ctx,
                interpreter,
            );
            const newCatchStatements = catchStatementsResult.stmts;
            ctx = catchStatementsResult.ctx;

            // Create the replacement node
            return {
                stmt: cloneAstNode({
                    ...stmt,
                    statements: newStatements,
                    catchStatements: newCatchStatements,
                }),
                ctx: ctx,
            };
        }
    }
}

function process_expression(
    expr: AstExpression,
    ctx: CompilerContext,
    interpreter: Interpreter,
): { expr: AstExpression; ctx: CompilerContext } {
    const value = tryExpressionSimplification(expr, interpreter);
    let newExpr = expr;
    if (value !== undefined) {
        try {
            newExpr = makeValueExpression(value, expr.loc);
            // Register the new expression in the context
            ctx = registerAllSubExpTypes(ctx, newExpr, getExpType(ctx, expr));
        } catch (_) {
            // This means that transforming the value into an AST node is
            // unsupported or it failed to register the type of the expression.
            // Just use the original expression.
            newExpr = expr;
        }
    }
    return { expr: newExpr, ctx: ctx };
}

function tryExpressionSimplification(
    expr: AstExpression,
    interpreter: Interpreter,
): Value | undefined {
    try {
        // Eventually, this will be replaced by the partial evaluator.
        return interpreter.interpretExpression(expr);
    } catch (e) {
        if (e instanceof TactConstEvalError) {
            if (!e.fatal) {
                return undefined;
            }
        }
        throw e;
    }
}

function registerAllSubExpTypes(ctx: CompilerContext, expr: AstValue, expType: TypeRef): CompilerContext {
    switch(expr.kind) {
        case "boolean":
        case "number":
        case "string":
        case "null": {
            ctx = registerExpType(ctx, expr, expType);
            break;
        }
        case "struct_instance": {
            ctx = registerExpType(ctx, expr, expType);

            const structFields = getType(ctx, expr.type).fields;
            const fieldTypes: Map<string, TypeRef> = new Map();
            
            for (const field of structFields) {
                fieldTypes.set(field.name, field.type);
            }

            for (const fieldValue of expr.args) {
                // Typechecking ensures that each field in the struct instance has a type
                const fieldType = fieldTypes.get(idText(fieldValue.field));
                if (fieldType === undefined) {
                    throwInternalCompilerError(`Field ${idText(fieldValue.field)} does not have a declared type in struct ${idText(expr.type)}.`, fieldValue.loc);
                }
                ctx = registerAllSubExpTypes(ctx, ensureAstValue(fieldValue.initializer, fieldValue.loc), fieldType);
            }
        }
    }
    return ctx;
}

function ensureAstValue(expr: AstExpression, src: SrcInfo): AstValue {
    switch(expr.kind) {
        case "boolean":
        case "null":
        case "number":
        case "string":
        case "struct_instance":
            return expr;
        default:
            throwInternalCompilerError(`Expressions of kind ${expr.kind} are not ASTValues.`, src);
    }
}
