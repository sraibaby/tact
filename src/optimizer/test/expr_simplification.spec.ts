import { featureEnable } from "../../config/features";
import { CompilerContext } from "../../context";
import { __DANGER_resetNodeId } from "../../grammar/ast";
import { openContext } from "../../grammar/store";
import { resolveDescriptors } from "../../types/resolveDescriptors";
import { getAllExpressionTypes } from "../../types/resolveExpression";
import { resolveStatements } from "../../types/resolveStatements";
import { loadCases } from "../../utils/loadCases";
import { simplify_expressions } from "../expr_simplification";

describe("expression-simplification", () => {
    beforeEach(() => {
        __DANGER_resetNodeId();
    });
    for (const r of loadCases(__dirname + "/success/")) {
        it("should pass expression simplification for " + r.name, () => {
            let ctx = openContext(
                new CompilerContext(),
                [{ code: r.code, path: "<unknown>", origin: "user" }],
                [],
            );
            ctx = featureEnable(ctx, "external");
            ctx = resolveDescriptors(ctx);
            ctx = resolveStatements(ctx);
            ctx = simplify_expressions(ctx);
            expect(getAllExpressionTypes(ctx)).toMatchSnapshot();
        });
    }
    for (const r of loadCases(__dirname + "/failed/")) {
        it("should fail expression simplification for " + r.name, () => {
            let ctx = openContext(
                new CompilerContext(),
                [{ code: r.code, path: "<unknown>", origin: "user" }],
                [],
            );
            ctx = featureEnable(ctx, "external");
            ctx = resolveDescriptors(ctx);
            ctx = resolveStatements(ctx);
            expect(() => {
                simplify_expressions(ctx);
            }).toThrowErrorMatchingSnapshot();
        });
    }
});
