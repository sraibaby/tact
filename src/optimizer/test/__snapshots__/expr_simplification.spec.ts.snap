// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expression-simplification should fail expression simplification for interpreter-called-when-no-contract 1`] = `
"<unknown>:6:9: Cannot evaluate expression to a constant: divisor expression must be non-zero
Line 6, col 9:
  5 | fun blowup(v: Int) {
> 6 |     1 / v;
              ^
  7 | }
"
`;
